[JSC] Inline Proxy IC calls in DFG

Constellation opened the pull request:
#### 22af8f1310b162dbcb59c0d76d60a6bfce24f67e
<pre>
[JSC] Inline Proxy IC calls in DFG
<a href="https://bugs.webkit.org/show_bug.cgi?id=276574">https://bugs.webkit.org/show_bug.cgi?id=276574</a>
<a href="https://rdar.apple.com/131675707">rdar://131675707</a>

Reviewed by Yijia Huang.

This patch inlines ProxyObject IC into DFG / FTL when IC site only see one ProxyObject IC.
As the same to the already-existing Proxy Load inlining in DFG, we do this for Store and In.
Also we support Indexed Proxy Load / Store / In in DFG inlining too.

                                           ToT                     Patched

    indexed-proxy-has-dfg-call       46.8285+-0.0983     ^     43.8616+-0.1500        ^ definitely 1.0676x faster
    indexed-proxy-get-dfg-call       49.7546+-0.0843     ^     45.7176+-0.2335        ^ definitely 1.0883x faster
    indexed-proxy-put-dfg-call       40.6735+-0.0854     ^     39.6159+-0.1099        ^ definitely 1.0267x faster

* JSTests/microbenchmarks/indexed-proxy-get-dfg-call.js: Added.
(test):
* JSTests/microbenchmarks/indexed-proxy-has-dfg-call.js: Added.
(test):
* JSTests/microbenchmarks/indexed-proxy-put-dfg-call.js: Added.
(test):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeForStubInfoWithoutExitSiteFeedback):
* Source/JavaScriptCore/bytecode/InByStatus.cpp:
(JSC::InByStatus::computeFor):
(JSC::InByStatus::computeForStubInfoWithoutExitSiteFeedback):
(JSC::InByStatus::merge):
(JSC::InByStatus::dump const):
(JSC::InByStatus::computeForStubInfo): Deleted.
* Source/JavaScriptCore/bytecode/InByStatus.h:
* Source/JavaScriptCore/bytecode/InByVariant.cpp:
(JSC::InByVariant::InByVariant):
(JSC::InByVariant::operator=):
(JSC::InByVariant::attemptToMerge):
(JSC::InByVariant::finalize):
(JSC::InByVariant::dumpInContext const):
* Source/JavaScriptCore/bytecode/InByVariant.h:
(JSC::InByVariant::callLinkStatus const):
* Source/JavaScriptCore/bytecode/InlineCallFrame.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
(JSC::InlineCallFrame::callModeFor):
(JSC::InlineCallFrame::specializationKindFor):
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::computeForStubInfo):
(JSC::PutByStatus::merge):
(JSC::PutByStatus::dump const):
* Source/JavaScriptCore/bytecode/PutByStatus.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::inlineCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectLoadCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectStoreCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectInCall):
(JSC::DFG::ByteCodeParser::handleProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handleProxyObjectStore):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectStore):
(JSC::DFG::ByteCodeParser::handleProxyObjectIn):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectIn):
(JSC::DFG::ByteCodeParser::handleInById):
(JSC::DFG::ByteCodeParser::handlePutById):
(JSC::DFG::ByteCodeParser::parseBlock):
(JSC::DFG::ByteCodeParser::handlePutByVal):
* Source/JavaScriptCore/dfg/DFGOSRExitCompilerCommon.cpp:
(JSC::DFG::callerReturnPC):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::returnLocationThunk):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::performProxyObjectHasFunctionConcurrently const):

Canonical link: <a href="https://commits.webkit.org/280973@main">https://commits.webkit.org/280973@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b686eb24aea07beaf84c3f775ea9c79a77176a95

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58114 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10594 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8559 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8750 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47084 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6099 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60144 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35106 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27918 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31869 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7535 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7563 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51206 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53816 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7803 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63440 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57356 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7863 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54359 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2035 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50251 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54457 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1734 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79117 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8682 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33271 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13142 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34357 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35441 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34102 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
```diff
@@ -3465,6 +3472,11 @@ llintOpWithMetadata(op_enumerator_get_by_val, OpEnumeratorGetByVal, macro (size,
     callSlowPath(_slow_path_enumerator_get_by_val)
     dispatch()
+
+.osrReturnPoint:
```
When we jump to this case? I didn't find the code for jumping to `.osrReturnPoint`.

Constellation replied:
This is used when OSR exit happens for the inlined function calls for this opcode. See returnLocation code.

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Inline Proxy IC calls in DFG
https://bugs.webkit.org/show_bug.cgi?id=276574
rdar://131675707

Reviewed by Yijia Huang.

This patch inlines ProxyObject IC into DFG / FTL when IC site only see one ProxyObject IC.
As the same to the already-existing Proxy Load inlining in DFG, we do this for Store and In.
Also we support Indexed Proxy Load / Store / In in DFG inlining too.

                                           ToT                     Patched

    indexed-proxy-has-dfg-call       46.8285+-0.0983     ^     43.8616+-0.1500        ^ definitely 1.0676x faster
    indexed-proxy-get-dfg-call       49.7546+-0.0843     ^     45.7176+-0.2335        ^ definitely 1.0883x faster
    indexed-proxy-put-dfg-call       40.6735+-0.0854     ^     39.6159+-0.1099        ^ definitely 1.0267x faster

* JSTests/microbenchmarks/indexed-proxy-get-dfg-call.js: Added.
(test):
* JSTests/microbenchmarks/indexed-proxy-has-dfg-call.js: Added.
(test):
* JSTests/microbenchmarks/indexed-proxy-put-dfg-call.js: Added.
(test):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeForStubInfoWithoutExitSiteFeedback):
* Source/JavaScriptCore/bytecode/InByStatus.cpp:
(JSC::InByStatus::computeFor):
(JSC::InByStatus::computeForStubInfoWithoutExitSiteFeedback):
(JSC::InByStatus::merge):
(JSC::InByStatus::dump const):
(JSC::InByStatus::computeForStubInfo): Deleted.
* Source/JavaScriptCore/bytecode/InByStatus.h:
* Source/JavaScriptCore/bytecode/InByVariant.cpp:
(JSC::InByVariant::InByVariant):
(JSC::InByVariant::operator=):
(JSC::InByVariant::attemptToMerge):
(JSC::InByVariant::finalize):
(JSC::InByVariant::dumpInContext const):
* Source/JavaScriptCore/bytecode/InByVariant.h:
(JSC::InByVariant::callLinkStatus const):
* Source/JavaScriptCore/bytecode/InlineCallFrame.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/InlineCallFrame.h:
(JSC::InlineCallFrame::callModeFor):
(JSC::InlineCallFrame::specializationKindFor):
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::computeForStubInfo):
(JSC::PutByStatus::merge):
(JSC::PutByStatus::dump const):
* Source/JavaScriptCore/bytecode/PutByStatus.h:
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::inlineCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectLoadCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectStoreCall):
(JSC::DFG::ByteCodeParser::emitProxyObjectInCall):
(JSC::DFG::ByteCodeParser::handleProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectLoad):
(JSC::DFG::ByteCodeParser::handleProxyObjectStore):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectStore):
(JSC::DFG::ByteCodeParser::handleProxyObjectIn):
(JSC::DFG::ByteCodeParser::handleIndexedProxyObjectIn):
(JSC::DFG::ByteCodeParser::handleInById):
(JSC::DFG::ByteCodeParser::handlePutById):
(JSC::DFG::ByteCodeParser::parseBlock):
(JSC::DFG::ByteCodeParser::handlePutByVal):
* Source/JavaScriptCore/dfg/DFGOSRExitCompilerCommon.cpp:
(JSC::DFG::callerReturnPC):
* Source/JavaScriptCore/llint/LLIntThunks.cpp:
(JSC::LLInt::returnLocationThunk):
* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter32_64.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter64.asm:
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::performProxyObjectHasFunctionConcurrently const):

Canonical link: https://commits.webkit.org/280973@main
